lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -2689,12 +2720,97 @@ void ModuleEmitter::emitHWGeneratedModule(HWModuleGeneratedOp module) {
-  auto instantiator = instance->getParentOfType<hw::HWModuleOp>().getName();
-  auto instanceName = instance.instanceName();
-  os << "bind " << instantiator << " " << instanceName << " " << instanceName
-     << " (.*);\n\n";
+// This may be called in the top-level, not just in an hw.module.  Thus we can't
+// use the name map to find expression names for arguments to the instance, nor
+// do we need to emit subexpressions.  Prepare pass, which has run for all
+// modules prior to this, has ensured that all arguments are bound to wires,
+// regs, or ports, with legalized names, so we can lookup up the names through
+// the IR.
+void ModuleEmitter::emitBind(BindOp op) {
+  InstanceOp inst = op.getReferencedInstance();
+
+  HWModuleOp parentMod = inst->getParentOfType<hw::HWModuleOp>();
+  auto parentVerilogName = getVerilogModuleNameAttr(parentMod);
+  verifyModuleName(op, parentVerilogName);
+
+  Operation *childMod = inst.getReferencedModule();
+  auto childVerilogName = getVerilogModuleNameAttr(childMod);
+  verifyModuleName(op, childVerilogName);
+
+  indent() << "bind " << parentVerilogName.getValue() << " "
+           << childVerilogName.getValue() << ' ' << inst.getName() << " (";
+
+  SmallVector<ModulePortInfo> parentPortInfo = parentMod.getPorts();
+  SmallVector<ModulePortInfo> childPortInfo = getModulePortInfo(childMod);
+
+  // Get the max port name length so we can align the '('.
+  size_t maxNameLength = 0;
+  for (auto &elt : childPortInfo) {
+    maxNameLength = std::max(maxNameLength, elt.getName().size());
+  }
+
+  // Emit the argument and result ports.
+  auto opArgs = inst.inputs();
+  auto opResults = inst.getResults();
+  bool isFirst = true; // True until we print a port.
+  for (auto &elt : childPortInfo) {
+    // Figure out which value we are emitting.
+    Value portVal = elt.isOutput() ? opResults[elt.argNum] : opArgs[elt.argNum];
+    bool isZeroWidth = isZeroBitType(elt.type);
+
+    // Decide if we should print a comma.  We can't do this if we're the first
+    // port or if all the subsequent ports are zero width.
+    if (!isFirst) {
+      bool shouldPrintComma = true;
+      if (isZeroWidth) {
+        shouldPrintComma = false;
+        for (size_t i = (&elt - childPortInfo.data()) + 1,
+                    e = childPortInfo.size();
+             i != e; ++i)
+          if (!isZeroBitType(childPortInfo[i].type)) {
+            shouldPrintComma = true;
+            break;
+          }
+      }
+
+      if (shouldPrintComma)
+        os << ',';
+    }
+    os << "\n";
+
+    // Emit the port's name.
+    indent();
+    if (!isZeroWidth) {
+      // If this is a real port we're printing, then it isn't the first one. Any
+      // subsequent ones will need a comma.
+      isFirst = false;
+      os << "  ";
+    } else {
+      // We comment out zero width ports, so their presence and initializer
+      // expressions are still emitted textually.
+      os << "//";
+    }
```

Comment:
This is duplicating logic from `visitStmt(InstanceOp op)`.  Is it possible to refactor these two together?

Reply:
There's overlap, but they diverge a fair amount.  I'll add an issue.